File: ueueo-framework/ueueo-uow/src/main/java/com/ueueo/uow/IUnitOfWork.java
package com.ueueo.uow;

import com.ueueo.ID;
import com.ueueo.IDisposable;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Predicate;

/**
 * TODO Description Of This JAVA Class.
 *
 * @author Lee
 * @date 2022-05-19 20:33
 */
public interface IUnitOfWork extends IDatabaseApiContainer, ITransactionApiContainer, IDisposable {
    ID getId();

    Map<String,Object> getItems();

    //TODO: Switch to OnFailed (sync) and OnDisposed (sync) methods to be compatible with OnCompleted
    BiConsumer<IUnitOfWork,UnitOfWorkFailedEventArgs> failed();

    BiConsumer<IUnitOfWork,UnitOfWorkEventArgs> disposed();

    IAbpUnitOfWorkOptions getOptions();

    IUnitOfWork getOuter();

    boolean  isReserved();

    boolean isDisposed();

    boolean isCompleted();

    String getReservationName();

    void setOuter(@Nullable IUnitOfWork outer);

    void initialize(@Nullable AbpUnitOfWorkOptions options);

    void reserve(@NonNull String reservationName);

    void saveChanges();

    void complete( );

    void rollback( );

    void onCompleted(Runnable handler);


    void addOrReplaceLocalEvent(
            UnitOfWorkEventRecord eventRecord,
            Predicate<UnitOfWorkEventRecord> replacementSelector
    );

    void addOrReplaceDistributedEvent(
            UnitOfWorkEventRecord eventRecord,
            Predicate<UnitOfWorkEventRecord> replacementSelector
    );
}

File: ueueo-framework/ueueo-uow/src/main/java/com/ueueo/uow/IUnitOfWork.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using JetBrains.Annotations;

namespace Volo.Abp.Uow;

public interface IUnitOfWork : IDatabaseApiContainer, ITransactionApiContainer, IDisposable
{
    ID Id;//  { get; }

    Dictionary<String, Object> Items;//  { get; }

    //TODO: Switch to OnFailed (sync) and OnDisposed (sync) methods to be compatible with OnCompleted
    event EventHandler<UnitOfWorkFailedEventArgs> Failed;

    event EventHandler<UnitOfWorkEventArgs> Disposed;

    IAbpUnitOfWorkOptions Options;//  { get; }

    IUnitOfWork Outer;//  { get; }

    boolean IsReserved;//  { get; }

    boolean IsDisposed;//  { get; }

    boolean IsCompleted;//  { get; }

    String ReservationName;//  { get; }

    void SetOuter(@Nullable IUnitOfWork outer);

    void Initialize(@Nonnull AbpUnitOfWorkOptions options);

    void Reserve(@Nonnull String reservationName);

    void SaveChangesAsync(CancellationToken cancellationToken = default);

    void CompleteAsync(CancellationToken cancellationToken = default);

    void RollbackAsync(CancellationToken cancellationToken = default);

    void OnCompleted(Func<Task> handler);

    void AddOrReplaceLocalEvent(
        UnitOfWorkEventRecord eventRecord,
        Predicate<UnitOfWorkEventRecord> replacementSelector = null
    );

    void AddOrReplaceDistributedEvent(
        UnitOfWorkEventRecord eventRecord,
        Predicate<UnitOfWorkEventRecord> replacementSelector = null
    );
}
